test/Microsoft.AspNetCore.Mvc.IntegrationTests/ModelBindingTestHelper.cs

```diff
@@ -136,23 +138,49 @@ private static IList<IModelValidatorProvider> GetModelValidatorProviders(IOption
         }
 
         private static HttpContext GetHttpContext(
+            IModelMetadataProvider metadataProvider,
             Action<HttpRequest> updateRequest = null,
-            Action<MvcOptions> updateOptions = null)
+            Action<MvcOptions> updateOptions = null,
+            MvcOptions mvcOptions = null)
         {
             var httpContext = new DefaultHttpContext();
             httpContext.Features.Set<IHttpRequestLifetimeFeature>(new CancellableRequestLifetimeFeature());
 
             updateRequest?.Invoke(httpContext.Request);
 
-            httpContext.RequestServices = GetServices(updateOptions);
+            httpContext.RequestServices = GetServices(metadataProvider, updateOptions, mvcOptions);
             return httpContext;
         }
 
-        public static IServiceProvider GetServices(Action<MvcOptions> updateOptions = null)
+        public static IServiceProvider GetServices(
+            IModelMetadataProvider metadataProvider,
+            Action<MvcOptions> updateOptions = null,
+            MvcOptions mvcOptions = null)
         {
             var serviceCollection = new ServiceCollection();
-            serviceCollection.AddAuthorization();
             serviceCollection.AddSingleton(new ApplicationPartManager());
+            if (metadataProvider != null)
+            {
+                serviceCollection.AddSingleton(metadataProvider);
+            }
+            else if (updateOptions != null || mvcOptions != null)
```

Comment:
I am a bit unclear about this. Should we depend on this check? Could we do a TryAddSingleton for this as we are always going to have options since we are doing AddMvc below?

Reply:
The check is about differentiating between cases where we should use `TestModelMetadataProvider.CreateDefaultProvider()` versus `CreateProvider(optionsAccessor.Value.ModelMetadataDetailsProviders)`. No need to check `MvcOptions` unless the calling test is changing those options i.e. might change `ModelMetadataDetailsProviders`.

Since these checks are about overriding the `IModelMetadataProvider` that `AddMvc()` would add, `TryAddSingleton(...)` would _not_ do the **Right Thing:tm:** for many tests.

Reply:
That's just a slight reordering of the current method. What advantage do you see?

Reply:
Just felt its better to read than what we currently have, but I will leave up to you.